action.yml

```diff
@@ -9,3 +21,73 @@ runs:
       uses: actions/setup-python@v4
       with:
         python-version: '3.9'
+
+    - name: "Config parser"
+      shell: bash
+      id: config-parser
+      env:
+        PUBLISH_CONFIG: ${{ inputs.publish-config }}
+        GITHUB_ACTION_PATH: ${{ github.action_path }}
+      run: |
+        mkdir -p ${{ inputs.temp-dir }}
+        python3 -m pip install pyyaml
+        python3 $GITHUB_ACTION_PATH/src/scripts/config_parser.py "${PUBLISH_CONFIG}"
+
+    - name: Checkout SVN
+      shell: bash
+      env:
+        repo_url: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.url }}
+      run: |
+        echo "Checking out SVN repo at $repo_url"
+        svn co $repo_url
+        echo "SVN repo checked out"
+      working-directory: "./${{ inputs.temp-dir }}"
+
+    - name: SVN check
+      shell: bash
+      if: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.svn-check.enabled == 'true' }}
+      env:
+        FILE_EXTENSIONS: ${{ toJson(fromJSON(steps.config-parser.outputs.pub_config).publishers.extensions) }}
+        PATH: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.path }}
+        NAME: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.svn-check.name }}
+        VERSION_FORMAT: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.version_pattern }}
+        SCRIPT: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.svn-check.script }}
+      working-directory: ./${{ inputs.temp-dir }}/${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.path }}
+      run: |
+        echo "Verifying $NAME"
+        python3 $SCRIPT
+
+    - name: Checksum check
+      shell: bash
+      if: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.checksum-check.enabled ==  'true' }}
+      env:
+        NAME: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.checksum-check.name }}
+        type: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.checksum-check.type }}
+        SCRIPT: ${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.rules.checksum-check.script }}
+      working-directory: ./${{ inputs.temp-dir }}/${{ fromJSON(steps.config-parser.outputs.pub_config).publishers.path }}
+      run: |
+        echo "Verifying $NAME"
+        chmod +x $SCRIPT
+        $SCRIPT "$type"
```

Comment:
I am not sure if I like this approach whith $SCRIPT. I think much better (and more composable) approach would be to have separate - very small actions for each of those scripts and compose this action out of smaller composable ones. This way we will not have to configure "scripts" in the configuration, but we would configure particular "actions" to be enabled. While it's less flexible, it would be a better abstraction IMHO. I somehow feel being able to provide a "script" to run is almost the same as writing your own action, where what we want, is to provide a user a set of predefined "actions" they can use. We do not want them to write new scripts.

Reply:
Ah, I see what you mean. That makes sense—using small, composable actions. So, I can split these steps into multiple actions within the same repository. What do you think?

As for the scripts I initially had in mind: the default script we provide would validate the rule. If someone wants to override it and use their custom script, they can specify the script location as a parameter in the config file. For example, in this workflow, the script is sourced from the workflow repository itself https://github.com/gopidesupavan/example-workspace/blob/main/publish-config.yml#L26.

On the other hand, for svn-check, no script is explicitly defined in the workflow https://github.com/gopidesupavan/example-workspace/blob/main/publish-config.yml#L18, so it falls back to the default script available in the action repository: https://github.com/gopidesupavan/gh-svn-pypi-publisher/blob/basic-feat/src/scripts/svn_checker.py.

Does this behavior okay to have?

Reply:
> Ah, I see what you mean. That makes sense—using small, composable actions. So, I can split these steps into multiple actions within the same repository. What do you think?

Yep.

> As for the scripts I initially had in mind: the default script we provide would validate the rule. If someone wants to override it and use their custom script, they can specify the script location as a parameter in the config file. For example, in this workflow, the script is sourced from the workflow repository itself https://github.com/gopidesupavan/example-workspace/blob/main/publish-config.yml#L26.

I see, but if we we have small composable actions, then it's even less effort to replace the action from "our" repo, with your own action you write - following the same pattern and writing your own hard-coded script. So in this case the users will get the "customizability" at the "sub-action" level, not the "script" level. Which I think is much better.

Reply:
Say (conceptually):

Our "composed" action:

```
Action "Publish to PyPI"
    Sub action verify SVN
    sub action verify Cheksum
    Sub action verify Signature
    Sub action Push To PyPI
```

And someone who would like to customize it, would copy our action to their repo and do this

```
Action "Publish to PyPI (my own)"
    Sub action verify SVN (my own)
    sub action verify Cheksum
    Sub action verify Signature
    Sub action Push To PyPI
```

Reply:
Ideally the "composed" action will be very simple and easy to copy and replace parts of it.

Reply:
Yes small composable actions will definitely  help here better way to handle these subsections and lot more easier for me to add this aswell :) great suggestion small composable action 🚀